R/differentiate-expr.R

```diff
@@ -0,0 +1,291 @@
+## multiplied by zero that should be removed.
+differentiate <- function(expr, name) {
+  if (is.numeric(expr)) {
+    0
+  } else if (is.symbol(expr)) {
+    if (identical(expr, as.symbol(name))) 1 else 0
+  } else {
+    fn <- as.character(expr[[1]])
+    if (!fn %in% names(derivative)) {
+      stop(sprintf("Unsupported function '%s' in differentiate()", fn))
+    }
+    derivative[[fn]](expr, name)
+  }
+}
+
+
+derivative <- list(
+  `+` = function(expr, name) {
+    maths$plus(differentiate(expr[[2]], name),
+               differentiate(expr[[3]], name))
+  },
+  `-` = function(expr, name) {
+    if (length(expr) == 3) {
+      maths$minus(differentiate(expr[[2]], name),
+                  differentiate(expr[[3]], name))
+    } else {
+      maths$uminus(differentiate(expr[[2]], name))
+    }
+  },
+  `*` = function(expr, name) {
+    a <- maths$rewrite(expr[[2]])
+    b <- maths$rewrite(expr[[3]])
+    da <- differentiate(a, name)
+    db <- differentiate(b, name)
+    maths$plus(maths$times(da, b), maths$times(a, db))
+  },
+  `/` = function(expr, name) {
+    a <- maths$rewrite(expr[[2]])
+    b <- maths$rewrite(expr[[3]])
+    da <- differentiate(a, name)
+    db <- differentiate(b, name)
+    maths$minus(
+      maths$divide(da, b),
+      maths$divide(maths$times(a, db), maths$times(b, b)))
+  },
+  `^` = function(expr, name) {
+    a <- maths$rewrite(expr[[2]])
+    b <- maths$rewrite(expr[[3]])
+    da <- differentiate(a, name)
+    db <- differentiate(b, name)
+    if (maths$is_zero(db)) {
+      maths$times(b, maths$times(maths$pow(a, maths$minus(b, 1)), da))
+    } else if (maths$is_zero(da)) {
+      maths$times(maths$pow(a, b), maths$times(call("log", a), db))
+    } else {
+      ## a^(b - 1) * (b da + a log(a) db)
+      maths$times(
+        maths$pow(a, maths$minus(b, 1)),
+        maths$plus(maths$times(b, da),
+                   maths$times(a, maths$times(call("log", a), db))))
+    }
+  },
+  `(` = function(expr, name) {
+    differentiate(expr[[2]], name)
+  },
+  exp = function(expr, name) {
+    a <- maths$rewrite(expr[[2]])
+    maths$times(differentiate(a, name), call("exp", a))
+  },
+  log = function(expr, name) {
+    a <- maths$rewrite(expr[[2]])
+    maths$divide(differentiate(a, name), a)
+  },
+  sqrt = function(expr, name) {
+    maths$divide(differentiate(expr[[2]], name),
+                 maths$times(2, maths$rewrite(expr)))
+  },
+  `if` = function(expr, name) {
+    condition <- maths$rewrite(expr[[2]])
+    da <- differentiate(expr[[3]], name)
+    db <- differentiate(expr[[4]], name)
+    if (identical(da, db)) da else call("if", condition, da, db)
+  },
+  lfactorial = function(expr, name) {
+    a <- maths$rewrite(expr[[2]])
+    maths$times(differentiate(a, name), call("digamma", maths$plus(a, 1)))
+  },
+  abs = function(expr, name) {
+    a <- maths$rewrite(expr[[2]])
+    maths$times(differentiate(a, name), call("sign", a))
+  }
+)
+
+maths <- local({
+  .protect <- function(x, except) {
```

Comment:
Up to you obviously but I think .bracket_except is clearer
```suggestion
  .bracket_except <- function(x, except) {
```

Reply:
yeah, this is nice. I've gone for `.parentheses_except` though as I believe:

* `(` parenthesis
* `[` bracket
* `{` brace

Reply:
ahh yes, i have unfortunately gotten into the bad habit of calling them:
* `(` bracket
* `[` square bracket
* `{` curly bracket

my bad :laughing: